usr/man/man4/hk.4
.TH HK 4 "27 July 1983"
.UC 4
.SH NAME
hk \- RK6-11/RK06 and RK07 moving head disk
.SH SYNOPSIS
.B "controller hk0 at uba? csr 0177440 vector rkintr"
.br
.B "disk rk0 at hk0 drive 0"
.SH DESCRIPTION
Files with minor device numbers 0 through 7 refer to various portions
of drive 0;
minor devices 8 through 15 refer to drive 1, etc.
The standard device names begin with ``hk'' followed by
the drive number and then a letter a-h for partitions 0-7 respectively.
The character ? stands here for a drive number in the range 0-7.
.PP
The block files access the disk via the system's normal
buffering mechanism and may be read and written without regard to
physical disk records.  There is also a `raw' interface
which provides for direct transmission between the disk
and the user's read or write buffer.
A single read or write call results in exactly one I/O operation
and therefore raw I/O is considerably more efficient when
many words are transmitted.  The names of the raw files
conventionally begin with an extra `r.'
.PP
In raw I/O counts should be a multiple of 512 bytes (a disk sector).
Likewise
.I seek
calls should specify a multiple of 512 bytes.
.SH "DISK SUPPORT"
The origin and size (in sectors) of the
pseudo-disks on each drive are as follows:
.PP
.nf
.ta .5i +\w'000000    'u +\w'000000    'u +\w'000000    'u
RK07 partitions:
	disk	start	length	cyl
	hk?a	0	15884	0-240
	hk?b	15906	10032	241-392
	hk?c	0	53790	0-814
	hk?g	26004	27786	393-813
.PP
RK06 partitions
	disk	start	length	cyl
	hk?a	0	15884	0-240
	hk?b	15906	11154	241-409
	hk?c	0	27126	0-410
.DT
.fi
.PP
On a dual RK-07 system
partition hk?a is used
for the root for one drive
and partition hk?g for the /usr file system.
If large jobs are to be run using
hk?b on both drives as swap area provides a 10Mbyte paging area.
Otherwise
partition hk?c on the other drive
is used as a single large file system.
.SH FILES
/dev/hk[0-7][a-h]	block files
.br
/dev/rhk[0-7][a-h]	raw files
.SH SEE ALSO
hp(4),
uda(4),
up(4)
.SH DIAGNOSTICS
\fBrk%d%c: hard error sn%d cs2=%b ds=%b er=%b\fR.  An unrecoverable
error occurred during transfer of the specified sector of the specified
disk partition.  The contents of the cs2, ds and er registers are printed
in octal and symbolically with bits decoded.
The error was either unrecoverable, or a large number of retry attempts
(including offset positioning and drive recalibration) could not
recover the error.
.PP
\fBrk%d: write locked\fR.  The write protect switch was set on the drive
when a write was attempted.  The write operation is not recoverable.
.PP
\fBrk%d: not ready\fR.  The drive was spun down or off line when it was
accessed.  The i/o operation is not recoverable.
.PP
\fBrk%d: not ready (came back!)\fR.  The drive was not ready, but after
printing the message about being not ready (which takes a fraction
of a second) was ready.  The operation is recovered if no further
errors occur.
.PP
\fBrk%d%c: soft ecc sn%d\fR.  A recoverable ECC error occurred on the
specified sector in the specified disk partition. 
This happens normally
a few times a week.  If it happens more frequently than
this the sectors where the errors are occurring should be checked to see
if certain cylinders on the pack, spots on the carriage of the drive
or heads are indicated.
.PP
\fBhk%d: lost interrupt\fR.  A timer watching the controller detected
no interrupt for an extended period while an operation was outstanding.
This indicates a hardware or software failure.  There is currently a
hardware/software problem with spinning down drives while they are
being accessed which causes this error to occur.
The error causes a UNIBUS reset, and retry of the pending operations.
If the controller continues to lose interrupts, this error will recur
a few seconds later.
.SH BUGS
In raw I/O
.I read
and
.IR write (2)
truncate file offsets to 512-byte block boundaries,
and
.I write
scribbles on the tail of incomplete blocks.
Thus,
in programs that are likely to access raw devices,
.I read, write
and
.IR lseek (2)
should always deal in 512-byte multiples.
.PP
DEC-standard error logging should be supported.
.PP
A program to analyze the logged error information (even in its
present reduced form) is needed.
.PP
The partition tables for the file systems should be read off of each
pack, as they are never quite what any single installation would prefer,
and this would make packs more portable.


named/CHANGES
This is version 4.8.1 of bind.  It is a test release that includes
version 4.8 with fixes, asynchronous zone transfer and better reload
capabilities.  Although it is currently described as a test release,
it is believed to be reasonably stable and more usable than the currently-
released version, 4.8.  The changes of note are:

  o The asynchronous zone transfer code previously posted to the bind
    mailing list has been integrated, completed and tested.  There are
    a number of changes from the version posted, including fixes to
    allow top-level domains to work and a simplification of the timer
    code.

  o The code for reloading the server has been changed so that only
    primary zones master files that have changed are reloaded.  The
    cache and secondary zones are not flushed, and the sortlist, domain,
    etc. are reset to correspond to the boot file contents.

  o Several bugs have been fixed: the name "*" is not interpreted as
    a wildcard in cached zones, only in primary zones.  Secondary servers
    no longer decrement the time-to-live of records by the time since
    they verified the zone with the master; as a result, they never
    hand out nameserver referrals with too short a TTL to be usable.
    A bug was fixed that caused secondary servers with out-of-date
    zones to return empty answers between the actual expiration time
    and the next timeout.

There are several other bugs that have been reported but have not yet
been fixed.  In addition, the next regular release of named will
support negative caching, but this has not been integrated.

I would appreciate receiving feedback on this release; in particular,
problems (or lack of problems) when installing on various systems.
I attempted to update the SysV code when integrating the zone-transfer,
but haven't tested it.

The notes from version 4.8 follow.

----------
Welcome to version 4.8 of bind.

There have been several changes to the named boot file (/etc/named.boot)
of which you should be aware.  The "domain" line for each zone is no longer
needed, but one such line may still be used to specify a default domain
to be used for queries containing names with only a single component.
The term "suffixes", which was added in version 4.7alpha, has been removed.

The manual page on named (named.8) has been updated to reflect all
these changes.  Please read this and look at the example files
before installation.  You should also note the changes in the
resolver code to support non-fully-qualified addresses and per-user
host aliases.  See hostname(7) for an overview.  Two new routines
have been added to the resolver library since the last test release:
res_query formulates a query, sends it, waits for a response and does
preliminary error checking; res_search implements the search rules
of gethostbyname using res_query.

The MX lookup routine in sendmail has been modified to use res_search.  
Also, dn_skip takes an additional parameter and has been renamed
to dn_skipname.  While old sendmail binaries will work with the new
version of bind, because of these changes, it is desirable to install
new sendmail sources and recompile sendmail.  Do not rebuild sendmail 
from old sources.   The new sendmail is on ucbarpa.Berkeley.EDU for
anonymous FTP from pub/4.3/sendmail.MX.tar and pub/4.3/sendmail.MX.tar.Z.

There have been numerous changes to named, fixing most of the known
bugs that can be fixed without major structural changes in the server.
Several server configurations that failed before should now work.
Certain robustness problems have been fixed, in particular bounds-
checking when processing incoming packets.  Two changes have been made
in preparation for negative caching: SOA records are sent in the authority
section in negative responses with NXDOMAIN set, and a bug was fixed that
caused confusion and repeated requests if a response had no error, no answer
and an SOA in the authority section.  As such responses are already sent
by other servers, and will be sent by the next release of BIND, it is
important that all sites upgrade to this version as quickly as possible.

The root "hint" cache and cache file remain the largest problem area,
along with named's naivete in accepting bogus server's data.
These will be addressed in the next release, along with asynchronous
zone transfers, intelligent reloading of zone files, faster startup,
and caching of negative responses.

This version (4.8) will replace the last officially released version (4.5).
Version 4.5 has a serious bug that causes the generation of a continuous
stream of bogons to the root domain servers (bogus queries with the query
response bit set and possibly garbage for nsid and rcode).  It is imperative
that these versions of named be replaced as fast as possible.  We urge you to
field 4.8 quickly, for the sake of the root domain servers.

	Mike Karels
	Jean Wood
	bind@ucbarpa.Berkeley.EDU


.ref-BSD-4/usr/src/sys/newdev/hk.c
static	char *sccsid = "@(#)hk.c	1.1 (Berkeley) 10/10/80";

/*
 * RK07 disk driver with bad-sector forwarding
 *
 *  This driver, revised from a UCSB RK07 driver, is modeled
 * after the Stanford SI 9500 driver with bad-sector forwarding.
 * The bad-sector information on the last track of each pack is
 * used to patch up transfers by forwarding accesses to bad sectors
 * to spares on the last cylinder.  The last cylinder is otherwise
 * write-protected.
 *  The bad-sector table is read when a minor device is first
 * opened after a close.  No Volume-Valid checking is done;
 * hamfists beware.  To be correct, the open routine needs a
 * mechanism for reading drive status and for doing a PACKACK.
 *  A pack is 1 minor device; there is no pack mapping.
 *  No attempt at ECC logic is included.  Be my guest.
 *
 *  BUGS:  The same bomb is planted here as in most UBA DMA drivers,
 * to wit, ubasetup is called from the interrupt level.  Ubasetup
 * will sleep if it cannot allocate; ka-boom!
 *
 * Author: Rob Mathews, EE Information Systems Lab, Stanford 4/80
 */

/*
 * Debug control
 *
 *  1 - open/close
 *  2 - strategy
 *  4 - intr entry/exit
 *  8 - intr detail
 * 16 - intr sector mapping
 * 32 - strat errors/lock-opens
 * 64 - bst reading
 *1024 - loop after errors
 */

#define debug(bit)	if (rk7debug & (bit)) printf
int rk7debug = (0);

#include "../h/param.h"
#include "../h/systm.h"
#include "../h/buf.h"
#include "../h/conf.h"
#include "../h/dir.h"
#include "../h/user.h"
#include "../h/map.h"
#include "../h/pte.h"
#include "../h/uba.h"

/*
 * Disk parameters
 */

#define	NRK7	8
#define NCYLS	815
#define NTRKS	3
#define NSECTS	22
#define NBYTES	512
#define	NBLKS	(NCYLS*NTRKS*NSECTS)
#define BSCYL	(NCYLS-1)		/* bad-sector repair cylinder */
#define BSBASE	(BSCYL*NTRKS*NSECTS)	/*  track 0,1 - repair pool */
#define BSTBLK	(BSBASE + 2*NSECTS)	/*  track 2 - bad-sector table */
#define MAXBB	10			/* max tolerable bad sectors */

/*
 * Register definitions
 */

#define	RK7ADDR	((struct rk7_regs *)(UBA0_DEV + 0177440))

struct	rk7_regs
{
	short rkcs1, rkwc;
	unsigned short rkba;
	short rkda, rkcs2, rkds, rker, rkasof, rkdc, rknull, rkdb, rkmr1;
	short rkecps, rkecpt, rkmr2, rkmr3;
};

/*
 * Register bits, per RK07 manual
 */

#define GO	01	/* cs1 */
#define DRCLR	04
#define	READ	020
#define	WRITE	022
#define	PACKACK	02
#define	IE	0100
#define	RDY	0200
#define BA1617	01400
#define CDT	02000
#define CERR	0100000
#define CCLR	0100000
#define SCLR	040	/* cs2 */
#define SVAL	0100000	/* ds */
#define DRDY	0200
#define VV	0100
#define DRA	01
#define DCK	0100000	/* er */
#define BSE	0200
#define NXF	04

#define Error	((RK7ADDR->rkcs1 & CERR) != 0)

/*
 * Driver parameters, data, and definitions
 */

#define RETRIES	10

#define IDLE	0
#define NORMAL	1
#define MAPPED	2
#define RESTART	3

struct buf rk7tab, rk7bsbuf, rrk7buf;

struct rk7
  {
	long access;
	short retries, errors, mapped, spurious;
	int ubinfo, rwcommand;
	short unsigned errer, errcs2, errds;
	short cylnow, trksecnow;
	unsigned short wcnow;
	long manow;
  }
 rk7;

struct bb
  {
	short serial, zeros[2], alignpack;
	struct bbtbl
	  {
		short bbtcyl, bbtts;
	  }
	 bbt[MAXBB];
  }
 rk7bb[NRK7];
#define OPENF	bbt[MAXBB-1].bbtcyl
#define LOCKF	zeros[0]

#define b_trksec av_back
#define b_cylin	 b_resid
#define nsects(x) x/NBYTES

/*
 */

rk7open (dev, flag)
  {
	register struct bb *bbp;
	register m;

	if ((m = minor (dev)) >= NRK7)
	  {
		u.u_error = ENXIO;
		return;
	  }
	bbp = &rk7bb[m];
	debug (1) ("open\n");
	if (bbp->OPENF != -1)
	  {
		bbp->OPENF = 1;
		debug (1|64) ("bs read\n");

		rk7bsbuf.b_flags = B_BUSY | B_READ;
		rk7bsbuf.b_dev = minor (dev);  /* major better not matter ! */
		rk7bsbuf.b_bcount = sizeof (struct bb);
		rk7bsbuf.b_un.b_addr = (caddr_t) bbp;
		rk7bsbuf.b_blkno = BSTBLK;
		rk7bsbuf.b_error = 0;
		rk7strategy (&rk7bsbuf);
		iowait (&rk7bsbuf);

		if (rk7bsbuf.b_flags & B_ERROR)
		  {
			printf ("rk7: error reading bad sector table!\n");
		  }
		 else if (bbp->OPENF != -1)
		  {
		    printf ("rk7: too many bad sectors, drive %d\n", m);
		    bbp->OPENF = -1;
		  }
		wakeup (&bbp->OPENF);
		debug (1|64) ("bs read done\n");
	  }
	 else
	  while (bbp->OPENF != -1)
	    sleep (&bbp->OPENF, PSWP);
  }

rk7close (dev, flag)
  {
	debug (1) ("close\n");
	if (!rk7bb[minor (dev)].LOCKF)
	  rk7bb[minor (dev)].OPENF = 0;
  }

rk7strategy (bp)
register struct buf *bp;
  {
	debug (2) ("strat %s ", bp->b_flags & B_READ ? "r" : "w");
	if (rk7bb[minor (bp->b_dev)].OPENF == 0)
	  {
		rk7open (bp->b_dev, 0);	/* in case root or swap is here */
		rk7bb[minor (bp->b_dev)].LOCKF++;
		debug (32) ("lock-open\n");
	  }

	if (bp->b_blkno + nsects (bp->b_bcount)
	      > (bp->b_flags & B_READ ? NBLKS : BSBASE))
	  {
		debug (2|32) ("error, blk %d, dev %d\n",
				bp->b_blkno, minor (bp->b_dev));
		bp->b_flags |= B_ERROR;
		iodone(bp);
		return;
	  }
	
	debug (2) ("queue ");
	bp->b_cylin = bp->b_blkno / (NTRKS*NSECTS);
	bp->b_trksec = (struct buf *) ((((bp->b_blkno / NSECTS) % NTRKS) << 8)
					+ (bp->b_blkno % NSECTS));
	
	spl5();
	disksort (&rk7tab, bp);

	if(rk7tab.b_active == IDLE)
	  rk7intr();
	spl0();
	debug (2) ("end\n");
  }

/*
disksort (tabp, bp)
register struct buf *tabp, bp;
  {
	bp->av_forw = (struct buf *)NULL;
	if(tabp->b_actf == NULL)
	  tabp->b_actf = bp;
	 else
	  tabp->b_actl->av_forw = bp;
	tabp->b_actl = bp;
  }
 */

/*
 * Interrupt (co)routine
 */

#define iwait(x)	rk7tab.b_active = x; return; case x:
rk7intr()
  {
   register struct buf *bp;

   bp = rk7tab.b_actf;
   switch (rk7tab.b_active)
    {
      case IDLE:
	debug (4) ("intr ");
	if (bp == NULL)
	  {
		rk7.spurious++;
		return;
	  }
	
	do { /* empty the queue */

		rk7.access++;
		rk7tab.b_errcnt = 0;

		do { /* transfer */
		     register short s;
		     short rk7map (), ms;

		     rk7.rwcommand = bp->b_flags & B_READ
					? (READ|IE|CDT|GO)
					: (WRITE|IE|CDT|GO);
		     if (Error)
		      {
			debug (8) ("clear ");
		        rk7clear ();
		      }

		     RK7ADDR->rkcs2 = minor (bp->b_dev);
		     RK7ADDR->rkcs1 = PACKACK|CDT|GO;
		     while (RK7ADDR->rkcs1 & GO);

		     debug (8) ("setup ");
		     rk7.ubinfo = ubasetup (bp, 1);
		     rk7start ( rk7.rwcommand,
				(short) bp->b_cylin, (short)(long) bp->b_trksec,
				(long) rk7.ubinfo,
				(short) (-(bp->b_bcount>>1)));
		     debug (8) ("started ");
		     iwait (NORMAL);

		     debug (8) ("NORMAL ");
		     while (RK7ADDR->rker == BSE && (ms = rk7map ()) != -1)
		      { /* forwarding */
			rk7.mapped++;
		     	debug (8|16) ("mapping ");

			rk7clear ();

			rk7start (rk7.rwcommand,
				  (short) BSCYL, (short) ms, (long) rk7.manow,
				  (short) (-(rk7.wcnow > NBYTES/2
						? NBYTES/2 : rk7.wcnow)));
			iwait (MAPPED);

		     	debug (8|16) ("mapped\n");
			if (!Error && rk7.wcnow > NBYTES/2)
			  {
				debug (8|16) ("restarting ");
				rk7restart ();
				iwait (RESTART);
				debug (8|16) ("restarted\n");
			  }
		      } /* forwarding */

		     if (Error)
		      {
			rk7.errer = RK7ADDR->rker;
			rk7.errcs2 = RK7ADDR->rkcs2;
			rk7.errds = RK7ADDR->rkds;
		      }
		     debug (8) (Error ? "error\n" : "free ");
		     ubafree (rk7.ubinfo);
		   } /* transfer */
		 while (Error && ++rk7tab.b_errcnt != RETRIES);

		if (rk7tab.b_errcnt != 0)
		  rk7errs (bp);

		debug (8) ("next\n");
		bp->b_resid = 0;
		rk7tab.b_actf = bp->av_forw;
		iodone (bp);

	   } /* empty the queue */
	 while ((bp = rk7tab.b_actf) != NULL);

	 rk7tab.b_active = IDLE;
	 debug (4) ("exit\n");
	 return;
    }
  }

rk7clear ()
  {
	register drive;

	drive = RK7ADDR->rkcs2 & 07;
	RK7ADDR->rkcs1 = CCLR;
	RK7ADDR->rkcs2 = drive;
	RK7ADDR->rkcs1 = DRCLR|CDT|GO;
	while ((RK7ADDR->rkcs1 & RDY) == 0);
  }

rk7start (rwcommand, cyl, tsect, mem, wc)
short cyl, tsect, wc;
long mem;
  {
	RK7ADDR->rkdc = cyl;
	RK7ADDR->rkda = tsect;
	RK7ADDR->rkba = mem;
	RK7ADDR->rkwc = wc;
	RK7ADDR->rkcs1 = ((mem >> 8) & BA1617) | rwcommand;
  }

short rk7map ()
  {
	register struct bbtbl *bbp;
	register i;

	rk7.wcnow = -RK7ADDR->rkwc;
	rk7.cylnow = RK7ADDR->rkdc;
	rk7.trksecnow = RK7ADDR->rkda;
	rk7.manow = (((long)(RK7ADDR->rkcs1 & BA1617)) << 8) + RK7ADDR->rkba;

	for (bbp = rk7bb[RK7ADDR->rkcs2 & 07].bbt, i = 0;
	     bbp->bbtcyl != -1 &&
	       (bbp->bbtcyl != rk7.cylnow || bbp->bbtts != rk7.trksecnow);
	     bbp++, i++);
	return (bbp->bbtcyl == -1 ? -1 : i /* trk & sec */);
  }

rk7restart ()
  {
	register short ts;

	ts = rk7.trksecnow;
	if ((++ts & 0377) == NSECTS)
	  if (((ts += (1<<8) - NSECTS) >> 8) == NTRKS)
	    {
		ts = 0;
		rk7.cylnow++;
	    }
	rk7start (rk7.rwcommand, (short) rk7.cylnow, (short) ts,
		  (long) rk7.manow + NBYTES, (short) (-(rk7.wcnow - NBYTES/2)));
  }

rk7errs (bp)
register struct buf *bp;
  {
	register nerr;

	nerr = rk7tab.b_errcnt;
	rk7.retries += nerr;
	if (nerr == RETRIES)
	  {
		bp->b_flags |= B_ERROR;
		rk7.errors++;
		printf ("Hard rk7 ");
	  }
	 else
	  printf ("%d * rk7 ", nerr);
	deverror (bp, rk7.errer, rk7.errcs2);
	printf ("ds %X\n", rk7.errds);
	while (nerr == RETRIES && (rk7debug & 1024));
  }

rk7read(dev)
dev_t dev;
{

	physio(rk7strategy, &rrk7buf, dev, B_READ, minphys);
}

rk7write(dev)
dev_t dev;
{

	physio(rk7strategy, &rrk7buf, dev, B_WRITE, minphys);
}
